document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/DocumentArrayFilterExpression.java

```diff
@@ -0,0 +1,68 @@
+package org.hypertrace.core.documentstore.expression.impl;
+
+import com.google.common.base.Preconditions;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Value;
+import org.hypertrace.core.documentstore.expression.operators.ArrayOperator;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.expression.type.SelectTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+
+/**
+ * Expression representing a condition for filtering on document array fields
+ *
+ * <p>Example: If product is an array field (containing documents)<code>
+ * ANY(product) [color IN ('Blue', 'Green') AND color != 'Black' AND name = 'Comb']
+ * </code> can be constructed as <code>
+ *   DocumentArrayFilterExpression.builder()
+ *    .operator(ANY)
+ *    .arraySource(IdentifierExpression.of("product"))
+ *    .filter(
+ *      LogicalExpression.and(
+ *        RelationalExpression.of(
+ *          IdentifierExpression.of("color"),
+ *          IN,
+ *          ConstantExpression.ofStrings("Blue", "Green")),
+ *        RelationalExpression.of(
+ *          IdentifierExpression.of("color"),
+ *          NEQ,
+ *          ConstantExpression.of("Black")),
+ *        RelationalExpression.of(
+ *          IdentifierExpression.of("name"),
+ *          EQ,
+ *          ConstantExpression.of("Comb"))
+ *      )
+ *    .build();
+ */
+@Value
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class DocumentArrayFilterExpression implements FilterTypeExpression {
+  ArrayOperator operator;
+
+  SelectTypeExpression arraySource;
```

Comment:
Q: Could this be anything else other than `IdentifierExpression`?

Reply:
For now, we only support the `IdentifierExpression`. Anything other than that will throw an exception from the `document-store` before going to the database. I deliberately made it `SelectTypeExpression` to support extensibility (E.g.: We want to apply filtering on some function that generates an array)